From 26ce8aa42914dd621e3f30d04bb531105352bac8 Mon Sep 17 00:00:00 2001
From: nmurugesan <nmurugesan@maxlinear.com>
Date: Thu, 22 Aug 2024 15:31:20 +0530
Subject: [PATCH] WLANRTSYS-84216 Reject Association request which has ML IE,
 without EHT capability

Issue:
UMI_STA_ADD is received in firmware with ML capability and without EHT
capability. FW asserted while setting rate for the station as the
rateMask is zero.

Fix:
Reject Association in hostapd if STA sends Assoc Req with ML IE
and without EHT capability.

FIXUP to WLANRTSYS-62845 MLO changes: MLO Auth Assoc, WPA3 and eapol
changes
---
 src/ap/ieee802_11.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index 8db141384..60c9e444e 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -5922,6 +5922,13 @@ static void handle_assoc(struct hostapd_data *hapd,
 	omit_rsnxe = !get_ie(pos, left, WLAN_EID_RSNX);
 
 #ifdef CONFIG_IEEE80211BE_MXL_MLO
+	/*If the assoc req has ML IE without EHT capability, reject the association.*/
+	if((sta->mxl_sta_info.ext_flags & WLAN_STA_MULTILINK) && !(sta->flags & WLAN_STA_EHT)) {
+		wpa_printf(MSG_DEBUG, "Reject ML STA " MACSTR " Assoc Request (missing EHT capability)", MAC2STR(mgmt->sa));
+		resp = WLAN_STATUS_UNSPECIFIED_FAILURE;
+		goto fail;
+	}
+
 	sta->mxl_sta_info.sta_profile_status = WLAN_STATUS_SUCCESS;
 	if (mxl_mlo_check_assoc_ie(hapd, sta, pos, left, reassoc, capab_info))
 		goto fail;
-- 
2.43.0

